### PR TITLE
[BEAM-9896] Added Snowflake streaming write 

### PR DESCRIPTION
**Pull request by kkucharc**

### CODE REVIEW

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 1]
```diff
@@ -169,7 +182,18 @@
...
   private static final String CSV_QUOTE_CHAR = "'";
-  private static final String WRITE_TMP_PATH = "data";
+
+  static final int DEFAULT_FLUSH_ROW_LIMIT = 10000;
+  static final int DEFAULT_STREAMING_SHARDS_NUMBER = 1;
+  static final int DEFAULT_BATCH_SHARDS_NUMBER = 0;
+  static final Duration DEFAULT_FLUSH_TIME_LIMIT = Duration.millis(30000); // 30 seconds
+  static final long SLEEP_STREAMING_LOGS_IN_SEC = 5000;
+  static final Duration STREAMING_LOGS_MAX_SLEEP = Duration.standardMinutes(2);;
+
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: Maybe is worth to make those constant names more consist ? 
DEFAULT_**STREAMING**_FLUSH_ROW_LIMIT
DEFAULT_STREAMING_SHARDS_NUMBER
DEFAULT_BATCH_SHARDS_NUMBER
DEFAULT__**STREAMING**_FLUSH_TIME_LIMIT
STREAMING_LOGS_SLEEP
STREAMING_LOGS_MAX_SLEEP

why STREAMING_LOGS_SLEEP is not using Duration?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 2]
```diff
@@ -169,7 +182,18 @@
...
+  public enum StreamingLogLevel {
+    INFO,
+    ERROR
+  }
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:193]: We have enums package. I wonder if we should keep this enum here or there.

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 3]
```diff
@@ -418,33 +455,58 @@ private String makeTmpDirName() {
...
       public void processElement(ProcessContext context) throws Exception {
 
-        SnowflakeServiceConfig config =
-            new SnowflakeServiceConfig(
-                dataSourceProviderFn, table, query, storageIntegrationName, stagingBucketDir);
+        String stagingBucketRunDir =
+            String.format(
+                "%s/%s/run_%s/",
+                stagingBucketDir, tmpDirName, UUID.randomUUID().toString().subSequence(0, 8));
+
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:499]: why subSequence 0,8 ?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 4]
```diff
@@ -477,14 +545,16 @@ public void processElement(ProcessContext context) throws Exception {
...
-      public CleanTmpFilesFromGcsFn(String stagingBucketDir) {
+      public CleanTmpFilesFromGcsFn(String stagingBucketDir, String tmpDirName) {
         this.stagingBucketDir = stagingBucketDir;
+        this.tmpDirName = tmpDirName;
       }
 
       @ProcessElement
       public void processElement(ProcessContext c) throws IOException {
-        String combinedPath = stagingBucketDir + "/**";
+        String combinedPath = stagingBucketDir + "/" + tmpDirName + "/**";
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: String.format?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java [hunk 5]
```diff
@@ -654,28 +774,40 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+      }
 
-      PCollection<String> out = write(input, stagingBucketDir);
       out.setCoder(StringUtf8Coder.of());
 
       return PDone.in(out.getPipeline());
     }
 
-    private void checkArguments() {
+    private void checkArguments(PCollection<T> input) {
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:813]: Maybe we can override `validate` method like here : https://github.com/apache/beam/blob/master/sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/CassandraIO.java#L1052

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java [hunk 6]
```diff
@@ -144,6 +155,11 @@ public void writeToExternalWithKVInput() {
                 .withSnowflakeService(snowflakeService));
 
     pipeline.run(options).waitUntilFinish();
+
+    List<String> actualData = FakeSnowflakeDatabase.getElements(FAKE_TABLE);
+    List<String> testDataInStrings =
+        testData.stream().map(e -> e.toString()).collect(Collectors.toList());
```
Comment (purbanow) [inline sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java]: Is it possible to make it like this : `Long::toString()` ?

**Comment by purbanow** [comment]
LGTM

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 1 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: Sure, I will change.

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 2 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:193]: Sure, let's do it

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 3 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:499]: To be honest I don't know. 8 chars seems random/long enough

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 4 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java]: Ok :)

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java (hunk 6 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java]: Let me check

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 3 above)
Comment (DariuszAniszewski) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:499]: That's correct - 8chars are enough for uniqueness in this case :)

**Review by DariuszAniszewski — APPROVED** [review]
👍

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java (hunk 6 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java]: It was poosible to change it:
` testData.stream().map(Object::toString).collect(Collectors.toList());`
Is it ok?

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 5 above)
Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:813]: Makes sense but then I am not sure how to check this check:
```
if (input.isBounded() == PCollection.IsBounded.UNBOUNDED) {
  checkArgument(getSnowPipe() != null, "withSnowPipe() is required");
}
```

Comment (kkucharc) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:813]: Ok, somehow this validation is not called before expand(). I would postpone this change.

File: sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java (hunk 6 above)
Comment (purbanow) [reply sdks/java/io/snowflake/src/test/java/org/apache/beam/sdk/io/snowflake/test/unit/write/SnowflakeIOWriteTest.java]: It is okey :D

File: sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java (hunk 5 above)
Comment (purbanow) [reply sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java:813]: Okie :)

**Pull request closed without merging**